src/System.Memory/src/System/ThrowHelper.cs

```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    //
+    // This pattern of inlinable "void Throw" routines that stack on top of NoInlining factory methods is a compromise between older JITs and newer JITs.
+    // The package this file appears in is targeted specifically at older JITs. Older JITs cannot inline methods that have conditional throws in them,
+    // thus we use these helpers to hide the throws. Newer JIT's will inline the Throw routines and will benefit from the code size reductions of
+    // splitting the creation details into their own non-inlinable factory methods.
+    //
```

Comment:
Comment should be something like
```csharp
//
// This pattern of easily inlinable "void Throw" routines that stack on top of NoInlining factory methods 
// is a compromise between older JITs and newer JITs (RyuJIT in Core CLR 1.1.0+ and desktop CLR in 4.6.3+).
//
// The aim of this pattern is three-fold
// 1. Extracting the throw makes the method preforming the throw in a conditional branch smaller and more inlinable
// 2. Extracting the throw from generic method to non-generic method reduces the repeated codegen size for value types
// 3a. Newer JITs will not inline the methods that only throw and also recognise them, move the call to cold section 
//     and not add stack prep and unwind before calling https://github.com/dotnet/coreclr/pull/6103
// 3b. Older JITs will inline the throw itself and move to cold section; but not inline the non-inlinable exception 
//     factory methods - still maintaining advantages 1 & 2
//
```
The conditional branches are in the methods that call the throws so currently unaffected by https://github.com/dotnet/coreclr/pull/8038